File: bench/crc32/crc32.c
#include "../_lib/bench.h"

#define N 16384
#define N_ITERS 220
#define N_RUNS 21
#define N_WARMUP 5

static uint8_t buf[N];
static uint32_t table[256];
static double samples[N_RUNS];

static void build_table(void) {
  for (uint32_t n = 0; n < 256; n++) {
    uint32_t c = n;
    for (int k = 0; k < 8; k++) c = (c & 1u) ? (0xedb88320u ^ (c >> 1)) : (c >> 1);
    table[n] = c;
  }
}

static void init_buf(void) {
  uint32_t x = 0x12345678u;
  for (int i = 0; i < N; i++) {
    x = x * 1103515245u + 12345u;
    buf[i] = (uint8_t)((x >> 16) & 0xffu);
  }
}

static uint32_t crc32_kernel(void) {
  uint32_t c = 0xffffffffu;
  for (int i = 0; i < N; i++) c = table[(c ^ buf[i]) & 0xffu] ^ (c >> 8);
  return c ^ 0xffffffffu;
}

static uint32_t run_kernel(void) {
  uint32_t h = 0;
  for (int it = 0; it < N_ITERS; it++) {
    h = mix_u32(h, crc32_kernel());
    int j = it % N;
    buf[j] = (uint8_t)((buf[j] + 1u) & 0xffu);
  }
  return h;
}

int main(void) {
  build_table();
  init_buf();
  uint32_t cs = 0;
  for (int i = 0; i < N_WARMUP; i++) cs = run_kernel();
  for (int i = 0; i < N_RUNS; i++) {
    double t0 = now_ms();
    cs = run_kernel();
    samples[i] = now_ms() - t0;
  }
  print_result(median_us(samples, N_RUNS), cs, N * N_ITERS, 1, N_RUNS);
}

File: bench/crc32/crc32.js
import { medianUs, mix, printResult } from '../_lib/benchlib.js'

// CRC-32 (IEEE, reflected) over a byte buffer, table-driven, hammered many
// passes per run. Pure-integer kernel: Uint8Array reads, an Int32Array LUT,
// `>>>`, `^`, `&` — all bit-exact between wasm i32 and JS, so jz and V8 must
// agree on the checksum. Exercises the typed-array-param + i32-narrowing path
// (`crc32(buf, table)` is the hot inner call) without any float arithmetic.

const N = 16384        // buffer length in bytes
const N_ITERS = 220    // CRC passes over the whole buffer per kernel run
const N_RUNS = 21
const N_WARMUP = 5

const buildTable = (table) => {
  for (let n = 0; n < 256; n++) {
    let c = n
    for (let k = 0; k < 8; k++) c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1)
    table[n] = c | 0
  }
}

const initBuf = (buf) => {
  let x = 0x12345678 | 0
  for (let i = 0; i < N; i++) {
    x = (Math.imul(x, 1103515245) + 12345) | 0
    buf[i] = (x >>> 16) & 0xff
  }
}

const crc32 = (buf, table) => {
  let c = 0xffffffff | 0
  for (let i = 0; i < N; i++) c = table[(c ^ buf[i]) & 0xff] ^ (c >>> 8)
  return (c ^ 0xffffffff) >>> 0
}

const runKernel = (buf, table) => {
  let h = 0
  for (let it = 0; it < N_ITERS; it++) {
    h = mix(h, crc32(buf, table) | 0)
    const j = it % N
    buf[j] = (buf[j] + 1) & 0xff   // perturb the buffer so the CRC can't be hoisted out of the loop
  }
  return h >>> 0
}

export let main = () => {
  const buf = new Uint8Array(N)
  const table = new Int32Array(256)
  buildTable(table)
  initBuf(buf)
  let cs = 0
  for (let i = 0; i < N_WARMUP; i++) cs = runKernel(buf, table)

  const samples = new Float64Array(N_RUNS)
  for (let i = 0; i < N_RUNS; i++) {
    const t0 = performance.now()
    cs = runKernel(buf, table)
    samples[i] = performance.now() - t0
  }
  printResult(medianUs(samples), cs, N * N_ITERS, 1, N_RUNS)
}
